xls/contrib/integrator/integration_algorithms/basic_integration_algorithm.cc

```diff
@@ -0,0 +1,104 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "xls/contrib/integrator/integration_algorithms/basic_integration_algorithm.h"
+
+#include "xls/ir/node_iterator.h"
+
+namespace xls {
+
+void BasicIntegrationAlgorithm::EnqueueNodeIfReady(Node* node) {
+  if (!queued_nodes_.contains(node) &&
+      integration_function_->AllOperandsHaveMapping(node)) {
+    ready_nodes_.push_back(node);
+    queued_nodes_.insert(node);
+  }
+}
+
+absl::Status BasicIntegrationAlgorithm::Initialize() {
+  // Make integration function.
+  XLS_ASSIGN_OR_RETURN(integration_function_, NewIntegrationFunction());
+
+  // ID initial nodes with all operands ready.
+  for (const Function* func : source_functions_) {
+    for (const Node* param : func->params()) {
+      for (Node* user : param->users()) {
+        EnqueueNodeIfReady(user);
```

Comment:
OOC, how does this deal with literals which have no operands? Do they get cloned over to the integration function?

Reply:
What will happen is that the algorithm will fail - good catch!  Fixed.

Interestingly, I expect my code at this point to two mux together two literals from different functions even if the literals are the same.  Yet, in a test case I added, I get 1 literal in the integrated function without the mux.  Are there some smarts somewhere besides the optimization passes (which I'm reasonably confident are being run) that would make this happen?